### PR TITLE
MAISTRA-2051: Add xns-informer for Kubernetes types

### PR DESCRIPTION
**Pull request by bison**
This is the first step in integrating the xns-informer library into the maistra-2.1 branch. This brings it into the vendor directory, and integrates it into the `pkg/kube` package, but only for core Kubernetes types and currently hard-coded to watch all namespaces. Integration with MemberRoll and other resource types will come later.

### CODE REVIEW

File: pkg/webhooks/validation/controller/controller_test.go [hunk 1]
```diff
@@ -257,6 +257,7 @@ func reconcileHelper(t *testing.T, c *fakeController) {
 }
 
 func TestGreenfield(t *testing.T) {
+	t.Skip("does not work with xns-informers currently")
```
Comment (jwendell) [inline pkg/webhooks/validation/controller/controller_test.go]: if this is not something you are going to fix in this pr, please create a jira for it

File: go.sum [hunk 2]
```diff
@@ -1049,6 +1052,7 @@ golang.org/x/crypto v0.0.0-20200220183623-bac4c82f6975/go.mod h1:LzIPMQfyMNhhGPh
 golang.org/x/crypto v0.0.0-20200414173820-0848c9571904 h1:bXoxMPcSLOq08zI3/c5dEBT6lE4eh+jOh886GHrn6V8=
 golang.org/x/crypto v0.0.0-20200414173820-0848c9571904/go.mod h1:LzIPMQfyMNhhGPhUkYOs5KpL4U8rLKemX1yGLhDgUto=
 golang.org/x/crypto v0.0.0-20200622213623-75b288015ac9/go.mod h1:LzIPMQfyMNhhGPhUkYOs5KpL4U8rLKemX1yGLhDgUto=
+golang.org/x/crypto v0.0.0-20200709230013-948cd5f35899/go.mod h1:LzIPMQfyMNhhGPhUkYOs5KpL4U8rLKemX1yGLhDgUto=
```
Comment (jwendell) [inline go.sum:1055]: worth to run a go mod tidy?

File: pkg/kube/client.go [hunk 3]
```diff
@@ -324,6 +339,14 @@ func newClientInternal(clientFactory util.Factory, revision string) (*client, er
 	}
 	c.dynamicInformer = dynamicinformer.NewDynamicSharedInformerFactory(c.dynamic, resyncInterval)
 
+	c.xnsInformerFactory = xnsinformers.NewSharedInformerFactoryWithOptions(
+		c.dynamic,
+		resyncInterval,
+		xnsinformers.WithNamespaces([]string{metav1.NamespaceAll}), // TODO: Pass namespaces.
```
Comment (jwendell) [inline pkg/kube/client.go:344]: Is this TODO going to be fixed in this PR?

File: go.sum (hunk 2 above)
Comment (bison) [reply go.sum:1055]: I did run `go mod tidy` -- does something look wrong here?

File: pkg/kube/client.go (hunk 3 above)
Comment (bison) [reply pkg/kube/client.go:344]: No, but I guess I was considering this part of the MAISTRA-2051 ticket. We can create a different one if you want though. Assuming this is about JIRA.

File: pkg/webhooks/validation/controller/controller_test.go (hunk 1 above)
Comment (bison) [reply pkg/webhooks/validation/controller/controller_test.go]: https://issues.redhat.com/browse/MAISTRA-2052

File: pkg/kube/client.go (hunk 3 above)
Comment (jwendell) [reply pkg/kube/client.go:344]: No strong opinion here, I'm just worried we could forget about it, so, better document it at some Jira

File: go.sum (hunk 2 above)
Comment (jwendell) [reply go.sum:1055]: nevermind then, I just saw many references to this crypto library and thought they needed a tidy.

**Comment by bison** [comment]
Everything passing, except the `TestMeshNetworking` test is now failing again here, but passes for me locally. Still not sure what's up with that one.

**Comment by dgn** [comment]
Looks like it was just a flake

**Review by dgn — APPROVED** [review]

File: pkg/webhooks/validation/controller/controller_test.go [hunk 4]
```diff
@@ -257,6 +257,9 @@ func reconcileHelper(t *testing.T, c *fakeController) {
 }
 
 func TestGreenfield(t *testing.T) {
+	// xns-informer issue: https://github.com/maistra/istio/issues/224
```
Comment (dgn) [inline pkg/webhooks/validation/controller/controller_test.go]: let's mention https://issues.redhat.com/browse/MAISTRA-2052 here

File: pkg/webhooks/validation/controller/controller_test.go [hunk 5]
```diff
@@ -299,6 +302,9 @@ func TestGreenfield(t *testing.T) {
 }
 
 func TestCABundleChange(t *testing.T) {
+	// xns-informer issue: https://github.com/maistra/istio/issues/224
```
Comment (dgn) [inline pkg/webhooks/validation/controller/controller_test.go]: same here, let's point to https://issues.redhat.com/browse/MAISTRA-2052

**Comment by bison** [comment]
Looks like this flake fails more than it passes. I've still had trouble reproducing it locally. Should we mark it as skipped for now so we can move on? I can keep looking into it in the meantime.

**Comment by dgn** [comment]
> Looks like this flake fails more than it passes. I've still had trouble reproducing it locally. Should we mark it as skipped for now so we can move on? I can keep looking into it in the meantime.

Yeah let's skip this for now and create a JIRA to track it

File: pkg/webhooks/validation/controller/controller_test.go (hunk 4 above)
Comment (bison) [reply pkg/webhooks/validation/controller/controller_test.go]: Fixed.

File: pkg/webhooks/validation/controller/controller_test.go (hunk 5 above)
Comment (bison) [reply pkg/webhooks/validation/controller/controller_test.go]: Fixed.

**Comment by bison** [comment]
Cool. I've marked it as skipped for now and linked to a JIRA issue. I'll see if I can figure out what's going on.

**Pull request merged**